### PR TITLE
fix fp16 layer norm error if epsilon is less than 1e-5

### PR DESCRIPTION
**Pull request by prathikr**
### Description

In mixed precision training setting, if epsilon is less than 1e-5 (default value) then downcasting from fp32 to fp16
sets the epsilon to 0. Consequently, the add node will be removed during round 1 of optimizations (since adding 0 is a no-op).
Thus, we account for this removal by making the add node check optional if performing mixed precision training.

### Motivation and Context

This fixes accuracy discrepancies seen during ort-enabled training where the training loss is on-par with baseline but the evaluation accuracy is far below baseline. See below screenshots for example of google/vit-base where we see the output of concat is fed into Add and then fused LayerNormalization in the training graph; however, in the evaluation graph it is fed to Add and then un-fused LayerNormalization subgraph.

vit_base_execution_model_training.onnx
image

vit_base_execution_model_inference.onnx
image

### CODE REVIEW

File: orttraining/orttraining/core/framework/ortmodule_graph_builder.cc [hunk 1]
```diff
@@ -76,15 +76,17 @@ Status OrtModuleGraphBuilder::Build(const std::vector<std::vector<int64_t>>* inp
     ORT_RETURN_IF_ERROR(SetConcreteInputShapes(*input_shapes_ptr));
   }
 
+  // Optimize the inference graph
+  std::unordered_set<std::string> x_node_arg_names;
+  ORT_RETURN_IF_ERROR(OptimizeForwardGraph(x_node_arg_names));
```
Comment (iK1D) [inline orttraining/orttraining/core/framework/ortmodule_graph_builder.cc:81]: We have two graph_transformer_utils.cc files, one is in onnxruntime folder (say, No.1) for any InferenceSession run, one is in orttraining folder (say, No.2) for training only, and it's applied to forward graph only.

For training, it's applying No.2 on forward graph, build gradient graph, and then apply No.1 during InferenceSession running
For inference, it's applying No.1 during InferenceSession running.

We cannot apply No.2 to eval mode because No.2 is for training only, some compute logic is different between training and eval (for example, BatchNorm). If you apply No.2 to eval mode, then during the eval mode, some Ops may run with logic that are for training because of some optimizers, which is not correct. That's the reason we but this OptimizeForwardGraph (which apply No.2 on forward graph) after the if and return. We cannot move it forward.

The LayerNormFusion is part of both No.1 and No.2, so for eval mode, logically the LayerNorm should also be fused when No.1 optimizers are applied. If not, then we need to check why. If it's because the structure of the subgraph is different (between training forward graph and eval graph) so that LayerNormFusion cannot fuse, then we need to check why the subgraph structure is different, or adjust the LayerNormFusion code to cover the new pattern.

**Comment by iK1D** [comment]
This is a known issue before. @mindest originally found this on one of the models year ago. I personally think it should be an issue from exporter side, it's not good for us to fix it in ORT (your fix actually change the semantic of the sub-graph, imagine that sub-graph is not exported, but created by user node by node, fuse it to LayerNorm with specific epsilon value change the compute result, which is not correct). Any code fix in ORT side actually hide bug from exporter.

When we logged the issue, I think exporter team actually had a fix (https://github.com/pytorch/pytorch/pull/89869) which adding Cast before and after the sub-graph to make sure the sub-graph is computed on float32 instead of float16. But from your case/screenshot here, I don't see the Cast nodes, not sure if you are using an old version of PyTorch. If not, I think it's better we ask exporter team to check and fix it, we may need to provide a minor reproduce. In the meantime, using a relatively larger epsilon value (within range of float16) by changing the modeling code directly could be a workaround for a short time.

Or if you check with some other people (inference guys, or Ashiwini, etc.) and they are fine with your solution, I am also fine with it (though I personally don't think it's a good one).
